scalanet/src/io/iohk/scalanet/peergroup/dynamictls/CustomHandlers.scala

```diff
@@ -0,0 +1,43 @@
+package io.iohk.scalanet.peergroup.dynamictls
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import io.netty.channel.ChannelHandler.Sharable
+import io.netty.channel.ChannelHandlerContext
+import io.netty.handler.ipfilter.AbstractRemoteAddressFilter
+
+import java.net.{InetAddress, InetSocketAddress}
+
+private[scalanet] object CustomHandlers {
+
+  /**
+    *
+    * Custom handler which keeps recent history of incoming connections. If it receive new connection from the ip address
+    * which is still in history, it rejects it as it means the remote caller tries to often.
+    *
+    * Handler needs to be thread safe as it is shared between several netty pipelines
+    *
+    * To share handlers between pipelines they need to be marked as @Sharable, if not netty refuses to share it.
+    */
+  @Sharable
+  class ThrottlingIpFilter(config: DynamicTLSPeerGroup.IncomingConnectionThrottlingConfig)
+      extends AbstractRemoteAddressFilter[InetSocketAddress] {
+
+    private val cacheView = Caffeine
+      .newBuilder()
+      .expireAfterWrite(config.throttlingDuration.length, config.throttlingDuration.unit)
+      .build[InetAddress, java.lang.Boolean]()
+      .asMap()
+
+    private def isRateLimited(address: InetAddress): Boolean = {
+      cacheView.putIfAbsent(address, java.lang.Boolean.TRUE) == null
```

Comment:
```suggestion
      cacheView.putIfAbsent(address, java.lang.Boolean.TRUE) != null
```
I thought if it wasn't null then it mean it was already in, so it has to be rate limited now.

Reply:
See my [other](https://github.com/input-output-hk/scalanet/pull/126#discussion_r570864482) comment about renaming the method to avoid double negation.